build: enable bundle_dts for core package

alan-agius4 opened the pull request:
`ng_module` will now include an `src/r3_symbol.d.ts` when compiling the core package under `ngc` togather with `dts bundling`, This is due that `ngcc` relies on this file to be present, but the `r3_symbols` file which is not part of our public api.

With this change, we can now ship an addition dts file which is flattened.

devversion commented on packages/compiler-cli/test/test_support.ts:
```diff
@@ -122,8 +122,15 @@ export function setupBazelTo(tmpDirPath: string) {
+    if (!fs.existsSync(path.join(angularDirectory, name, 'index.d.ts'))) {
+      fs.symlinkSync(
+          path.join(pkgPath, `${name}.d.ts`), path.join(angularDirectory, name, 'index.d.ts'));
+    }
```
What's the state of that change? should that go in as well? I don't see why this is necessary. Can you add some context?

alan-agius4 replied:
I haven’t figure this out yet, but a couple of tests, won’t resolve @angular/core typings based on path in package.json, but it seems it’s being resolved using an index.d.ts.

devversion replied:
Doesn't that mean that something is wrong with either how the test environment consumes the typings or how they are declared in the `package.json`?

alan-agius4 replied:
it’s the test setup. I found a couple of other issues with the test setup.

See: https://github.com/angular/angular/commit/58b40453597b1d03353b046fe3d1a3cedef08b1d

This causing missing module resolutions. 

Similar to this error here. But for this one I haven’t look into remove this work around here yet.

alan-agius4 replied:
Probably we are skipping to replace an implementation of a method or something.

devversion replied:
Yeah. we should look into that (not needed for this PR though, I realize) 👍

alan-agius4 replied:
It’s in my todo

alexeagle commented:
[presubmit](http://test/OCL:235552862:BASE:235552947:1551118566992:8e7e630f)

alan-agius4 commented:
Seems like it’s failing in G3

IgorMinar commented on packages/bazel/test/ng_package/example/secondary/additional.ts:
```diff
@@ -0,0 +1,9 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+export * from './additionalmodule';
```
please rename the file `additiona_module` to stick with naming conventions?

alan-agius4 replied:
Will address this, when we decide on the below.

IgorMinar commented on packages/core/BUILD.bazel:
```diff
@@ -10,6 +10,10 @@ ng_module(
+    additional_dts_bundles = [
+        "src/r3_symbols.d.ts",
+    ],
+    bundle_dts = True,
```
can we make this the default for all public facing ng_modules? it seems error prone to keep on opting in.

alan-agius4 replied:
Now we can as a lot of the refactoring is already done, however this PR will grow a bit, since I have another 3 PRs open to enable `bundle_dts` in another packages.

@angular/common (https://github.com/angular/angular/pull/28834)
@angular/forms (https://github.com/angular/angular/pull/28854)
@angular/router (https://github.com/angular/angular/pull/28833)

However, I can port the changes into this PR, is you prefer to do so.

Alternatively, I can change the default value of flag, when all PRs get merged.

alan-agius4 replied:
Separate Pr for this https://github.com/angular/angular/pull/29128

IgorMinar reviewed (dismissed):
hmm.. could we make this a private feature? the reason we need this at all is because we live in a dual world between view engine and ivy. otherwise we wouldn't need it, and I don't feel comfortable exposing this to the world and being on the hook to maintain it.

if anything `additional_dts_bundles` is not the right name for this feature because as long as I understand it correctly, it doesn't produce additional bundles, this is just a list of additional entry points that need to be pulled into the dts bundle already being produced.

in any case, this feature seems very very specific to a hack we need only temporarily while working on Ivy. Can we do better here?

alan-agius4 commented:
When using `additional_dts_bundles` it will actually produce a separate dts bundle. In the case of `core` it will produce `core.d.ts` and `r3_symbols.d.ts`, the later not being referenced in the `typings` field.

I can think of some options if we'd don't want to make this private.
1) prefix the attribute with an `_`
2) Internally within `ng_module` add the `r3_symbols` bundle if the build target module name is `@angular/core`.

What's your preferred approach?

IgorMinar commented:
Huh? Something doesn't add up here. Do we have a runtime version of that bundle that I'm not aware of?

alan-agius4 commented:
Ngcc uses this https://github.com/angular/angular/blob/3842dd6a6d342ce993f418920b9bf87128ae63c1/packages/compiler-cli/src/ngcc/src/packages/entry_point_bundle.ts#L54

IgorMinar commented:
I suggest we do a quick sync with @alxhub and @petebacondarwin to make sure we are doing the right thing here.. maybe we are, but it seems that not enough people are aware of what's going on here, so there is a risk that we are building the wrong thing.

alexeagle commented:
discussing in #dts-bundling channel on slack

alexeagle commented on packages/bazel/src/ng_module.bzl:
```diff
@@ -687,6 +690,16 @@ NG_MODULE_RULE_ATTRS = dict(dict(COMMON_ATTRIBUTES, **NG_MODULE_ATTRIBUTES), **{
     # https://github.com/angular/angular/blob/master/packages/compiler-cli/src/transformers/api.ts
     "flat_module_out_file": attr.string(),
     "bundle_dts": attr.bool(default = False),
+    "additional_dts_bundles": attr.string_list(
```
if you're confident that `@angular/core` is the only `ng_module` rule that would ever need this attribute, then let's just hard-code the behavior to apply specifically to that package and not add an attribute.
Because we'll just remove this attribute once ngcc is fixed to not require the r3_symbols bundle.

alan-agius4 replied:
Adrrsssed

alan-agius4 committed:
build: enable bundle_dts for core package

`ng_module` will now include an `src/r3_symbol.d.ts` when compiling the core package under `ngc` togather with `dts bundling`, This is due that `ngcc` relies on this file to be present, but the `r3_symbols` file which is not part of our public api.

With this change, we can now ship an addition dts file which is flattened.

alan-agius4 committed:
test: fix ngcc tests to work with flattened dts files

alan-agius4 committed:
test: add interm fix for test that rely on an index.d.ts file

At the moment, certain tests relies on resolving the module with an index.d.ts, this root cause might be some implementations are missing from the mocks.

Similar to: https://github.com/angular/angular/commit/58b40453597b1d03353b046fe3d1a3cedef08b1d
